### PR TITLE
Adding static html support as a coverage service

### PR DESCRIPTION
**Pull request by ikhsan**
I'm fairly novice on ruby and slather's codebase, so if there are codes that need to be improved or probably a guideline that I need to conform to, please inform me.

This will generate static html files, so reviewing code locally can be more thorough than using the simple output. Often the simple output is enough, but you only know __how much__ lines instead of __which__ lines are being covered.

This is useful for me because it will just open in the browser and it doesn't need any connection or application (no need coverstory mentioned in #49).

Here's what it looks like :tada:

slather_html_file
slather_html_index

### CODE REVIEW

**Comment by ixnixnixn** [comment]
One thing to notice, although the [build is passing](https://travis-ci.org/ixnixnixn/slather/builds/61022696), coveralls sends an error message:
```
Coveralls encountered an exception:
Errno::EMFILE
Too many open files - getcwd
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/configuration.rb:116:in `pwd'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/configuration.rb:116:in `pwd'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/configuration.rb:171:in `relevant_env'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/configuration.rb:9:in `configuration'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/api.rb:64:in `apified_hash'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/api.rb:21:in `post_json'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/coveralls-0.8.1/lib/coveralls/simplecov.rb:72:in `format'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/simplecov-0.10.0/lib/simplecov/result.rb:46:in `format!'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/simplecov-0.10.0/lib/simplecov/configuration.rb:159:in `block in at_exit'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/simplecov-0.10.0/lib/simplecov/defaults.rb:60:in `call'
/Users/travis/build/ixnixnixn/slather/travis_bundle_dir/ruby/2.0.0/gems/simplecov-0.10.0/lib/simplecov/defaults.rb:60:in `block in <top (required)>'
```

I've tried to close all the files in my specs but it doesn't remove the error message.

File: css/slather.css [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/* --------------------------------------------------------
```
Comment (ixnixnixn) [inline css/slather.css]: I don't know where to put the css file, so I put it in the root folder. And all of the generated html will link to this css from the gem's root folder.

I grab the colour scheme from the logo, and 'slather' it to the stylesheet whenever appropriate. I've used coverall's HTML/CSS as a reference.

File: lib/slather/coverage_service/html_output.rb [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+        FileUtils.rm_rf(directory_path) if File.exists?(directory_path)
+        FileUtils.mkdir_p(directory_path)
+
+        reports.each do |name, doc|
+          html_file = File.join(directory_path, "#{name}.html")
+          File.write(html_file, doc.to_html)
+        end
+
+        # We could open 'index.html' automatically, although I don't know how to disable it for testing
```
Comment (ixnixnixn) [inline lib/slather/coverage_service/html_output.rb]: We could open the index html automatically, but is that appropriate? or should we just leave the user to decide when to open it.

File: spec/slather/coverage_service/html_output_spec.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+      def extract_coverage_text(doc)
+        doc.at_css('#total_coverage').text
+      end
+
+      def extract_coverage_class(doc)
+        doc.at_css('#total_coverage').attribute("class").to_s
+      end
+
+      def extract_cov_index(doc)
```
Comment (ixnixnixn) [inline spec/slather/coverage_service/html_output_spec.rb]: I don't know how to properly compare two html files so I opted for grabbing the information that matters with nokogiri. If there's a better way (or probably a gem?) to compare html files, then please let me know.

**Comment by marklarr** [comment]
Thanks! I'll take a look at this

**Comment by marklarr** [comment]
This is super cool.

**Comment by ixnixnixn** [comment]
Thanks! Please inform me if there's anything that I can improve.

**Comment by Ashton-W** [comment]
This is awesome @ixnixnixn :+1:

**Comment by ixnixnixn** [comment]
@Ashton-W thanks mate! Although I haven't put any js stuff for sorting and filtering.

**Comment by Ashton-W** [comment]
Might be best to get this merged before adding that - I'm keen to help.

**Comment by 3lvis** [comment]
OMG. can't wait to be able to use this!!!

**Comment by ixnixnixn** [comment]
@Ashton-W @3lvis probably there are more experienced rubyist in this thread, but do you know how to rectify the [exception from coveralls](https://travis-ci.org/venmo/slather/builds/61023477)? It's currently passing but coveralls seemed to be complaining :confused: 
```
Coveralls encountered an exception:
Errno::EMFILE
Too many open files - getcwd
```

**Comment by tarbrain** [comment]
Also waiting 🚬

**Comment by seanoshea** [comment]
Really cool. Would love to see this merged :thumbsup:

File: lib/slather/coverage_service/html_output.rb (hunk 2 above)
Comment (marklarr) [inline lib/slather/coverage_service/html_output.rb]: let's open it automatically... you could just make a method, and then stub it out in rspec

```
def open(index_html)
  `open #{index_html}` if File.exists?(index_html)
end

#spec
project.stub(:open)

```

Comment (Ashton-W) [inline lib/slather/coverage_service/html_output.rb]: We could print a file URL, most terminals allow 'right click to open URL'. And it wouldn't cause trouble on CI servers.

**Comment by ixnixnixn** [comment]
Updates! :tada: 
* after slathed it'll open index.html automatically
* no more `Errno::EMFILE Too many open files - getcwd` after updating to ruby 2.2
* syntax highlighting with [highlight.js](https://highlightjs.org)
syntax-highlighting

Tests are passing, everything should be sweet :+1:

**Comment by Ashton-W** [comment]
I think opening the browser should be tied to an option flag, CI servers need a way to stop it.

**Comment by seanoshea** [comment]
FWIW, https://github.com/jonreid/XcodeCoverage allows this to be configured via a `--show` flag.

**Comment by tarbrain** [comment]
Favoring the '--show' flag as well. On a CI server you only need the html output for publishing.

**Comment by ixnixnixn** [comment]
Hmm, fair point. I do want to open it by default because I was thinking that this will only be for local usage. Hence, I missed the CI related issues. 

My plan will be printing the path by default, and if `--show` is flagged then it will open the browser. I'll give it a go tonight.

**Comment by ixnixnixn** [comment]
Another update 🎉
* using `--show` flag for opening "`index.html`" automatically, otherwise print out path by default
* coverage list supports sorting and filtering via [List.js](http://www.listjs.com)

screen shot 2015-06-05 at 15 36 12

**Comment by marklarr** [comment]
:+1: thanks @ikhsan and everyone else.

**Comment by ikhsan** [comment]
Yay! Thanks Mark and all! :tada:

**Comment by ikhsan** [comment]
Anyways, I've filed another minor PR to update the docs related to the html support (removing -h, updating flags, and my link) : #90

**Pull request merged**